jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java

```diff
@@ -159,6 +170,419 @@ default void clear()
         BufferUtil.clear(getByteBuffer());
     }
 
+    default int append(byte[] bytes, int offset, int length)
+    {
+        ByteBuffer to = getByteBuffer();
+
+        int pos = BufferUtil.flipToFill(to);
+        try
+        {
+            length = Math.min(length, to.remaining());
+            to.put(bytes, offset, length);
+        }
+        finally
+        {
+            BufferUtil.flipToFlush(to, pos);
+        }
+        return length;
+    }
+
+    default void append(ByteBuffer bytes)
+    {
+        BufferUtil.append(getByteBuffer(), bytes);
+    }
+
+    default boolean append(ByteBuffer bytes, Runnable releaser)
+    {
+        append(bytes);
+        if (bytes.hasRemaining())
+            return false;
+        releaser.run();
+        return true;
+    }
+
+    default void append(RetainableByteBuffer bytes)
+    {
+        bytes.writeTo(this);
+    }
+
+    default void putTo(ByteBuffer toInfillMode)
+    {
+        ByteBuffer slice = getByteBuffer().slice();
+        toInfillMode.put(slice);
+    }
+
+    default boolean writeTo(ByteBuffer to)
+    {
+        ByteBuffer slice = getByteBuffer().slice();
+        BufferUtil.append(to, slice);
+        return slice.remaining() == 0;
+    }
+
+    default boolean writeTo(RetainableByteBuffer to)
+    {
+        if (remaining() == 0)
+            return true;
+        ByteBuffer slice = getByteBuffer().slice();
+        to.append(slice);
+        return slice.remaining() == 0;
+    }
+
+    default void writeTo(OutputStream out) throws IOException
```

Comment:
I'd rather not have a variant that writes to an `OutputStream` unless needed, and so far, only tests use this method.

Reply:
The whole point is so that we can efficiently write accumulated buffers to something like an OutputStream.  I think it has to be there, since when it is needed an efficient version is unlikely to be written and getByteBuffer will just be called.

Reply:
Wouldn't a `Content.Sink` to `OutputStream` adapter like the following give the same benefits without needed extra API?

```java
public class OutputStreamContentSink implements Content.Sink
{
    private final OutputStream outputStream;
    private final int tempBufferSize;
    private byte[] tempBuffer;
    private boolean lastWritten;

    public OutputStreamContentSink(OutputStream outputStream)
    {
        this(outputStream, 1024);
    }

    public OutputStreamContentSink(OutputStream outputStream, int tempBufferSize)
    {
        this.outputStream = outputStream;
        this.tempBufferSize = tempBufferSize;
    }

    @Override
    public void write(boolean last, ByteBuffer byteBuffer, Callback callback)
    {
        if (lastWritten)
        {
            callback.failed(new IOException("complete"));
            return;
        }
        lastWritten = last;

        Throwable failure = null;
        try
        {
            if (byteBuffer.hasArray())
            {
                outputStream.write(byteBuffer.array(), byteBuffer.arrayOffset() + byteBuffer.position(), byteBuffer.remaining());
                byteBuffer.position(byteBuffer.position() + byteBuffer.remaining());
            }
            else
            {
                if (tempBuffer == null)
                    tempBuffer = new byte[tempBufferSize];
                while (byteBuffer.hasRemaining())
                {
                    int byteCountToWrite = Math.min(byteBuffer.remaining(), tempBuffer.length);
                    byteBuffer.get(tempBuffer, 0, byteCountToWrite);
                    outputStream.write(tempBuffer, 0, byteCountToWrite);
                }
            }

            callback.succeeded();
        }
        catch (Throwable x)
        {
            failure = x;
        }

        try
        {
            if (last)
            {
                tempBuffer = null;
                outputStream.close();
            }
        }
        catch (Throwable x)
        {
            if (failure == null)
                failure = x;
            else
                ExceptionUtil.addSuppressedIfNotAssociated(failure, x);
        }

        if (failure != null)
        {
            tempBuffer = null;
            callback.failed(failure);
        }
    }
}
```

Reply:
Hmmmm the PR has 1 new method, with a 1 line default implementation and a 2 line optimized implementation for aggregation.

This is a whole new class with a over 60 lines of impl (although it could be shortened to about 30 I think if it used BufferUtil.writeTo).

I'm not opposed to having such a class... but I just don't see it as simpler than the writeTo method.   Let me have a go of writing the simplest impl I can and then we can decide....

Reply:
OK, I was able to implement that with an 11 line anonymous lambda.   So I have removed this writeTo method.